liquidity/fees.go

```diff
@@ -0,0 +1,202 @@
+package liquidity
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/loop"
+)
+
+const (
+	// defaultSwapFeePPM is the default limit we place on swap fees,
+	// expressed as parts per million of swap volume, 0.5%.
+	defaultSwapFeePPM = 5000
+
+	// defaultRoutingFeePPM is the default limit we place on routing fees
+	// for the swap invoice, expressed as parts per million of swap volume,
+	// 1%.
+	defaultRoutingFeePPM = 10000
+
+	// defaultRoutingFeePPM is the default limit we place on routing fees
+	// for the prepay invoice, expressed as parts per million of prepay
+	// volume, 0.5%.
+	defaultPrepayRoutingFeePPM = 5000
+
+	// defaultMaximumMinerFee is the default limit we place on miner fees
+	// per swap. We apply a multiplier to this default fee to guard against
+	// the case where we have broadcast the preimage, then fees spike and
+	// we need to sweep the preimage.
+	defaultMaximumMinerFee = 15000 * 100
+
+	// defaultMaximumPrepay is the default limit we place on prepay
+	// invoices.
+	defaultMaximumPrepay = 30000
+)
+
+var (
+	// ErrZeroMinerFee is returned if a zero maximum miner fee is set.
+	ErrZeroMinerFee = errors.New("maximum miner fee must be non-zero")
+
+	// ErrZeroSwapFeePPM is returned if a zero server fee ppm is set.
+	ErrZeroSwapFeePPM = errors.New("swap fee PPM must be non-zero")
+
+	// ErrZeroRoutingPPM is returned if a zero routing fee ppm is set.
+	ErrZeroRoutingPPM = errors.New("routing fee PPM must be non-zero")
+
+	// ErrZeroPrepayPPM is returned if a zero prepay routing fee ppm is set.
+	ErrZeroPrepayPPM = errors.New("prepay routing fee PPM must be non-zero")
+
+	// ErrZeroPrepay is returned if a zero maximum prepay is set.
+	ErrZeroPrepay = errors.New("maximum prepay must be non-zero")
+)
+
+// FeeLimit is an interface implemented by different strategies for limiting
+// the fees we pay for autoloops.
+type FeeLimit interface {
+	// String returns the string representation of fee limits.
+	String() string
+
+	// validate returns an error if the values provided are invalid.
+	validate() error
+
+	// loopOutLimits checks whether the quote provided is within our fee
+	// limits for the swap amount.
+	loopOutLimits(amount btcutil.Amount, quote *loop.LoopOutQuote) error
+
+	// loopOutFees return the maximum prepay and invoice routing fees for
+	// a swap amount and quote.
+	loopOutFees(amount btcutil.Amount, quote *loop.LoopOutQuote) (
+		btcutil.Amount, btcutil.Amount, btcutil.Amount)
+}
+
+// FeeCategoryLimit is an implementation of the fee limit interface which sets
+// a specific fee limit per fee category.
+type FeeCategoryLimit struct {
+	// MaximumPrepay is the maximum prepay amount we are willing to pay per
+	// swap.
+	MaximumPrepay btcutil.Amount
+
+	// MaximumSwapFeePPM is the maximum server fee we are willing to pay per
+	// swap expressed as parts per million of the swap volume.
+	MaximumSwapFeePPM int
```

Comment:
Define as `uint64` to make fewer conversions necessary? Same with the constants above?